File: test/Principal.py
import unittest as ut
import numpy as np
import Maniverse as mv

# Principal component analysis
# Finding the space spanned by the highest 5 eigenvectors
# Maximize L(C) = Tr[ C.t A C ]
# A \in Sym(10)
# C \in Flag(5; 10) = Gr(5; 10)

class Obj(mv.Objective):
	def __init__(self):
		super().__init__()
		self.A = np.loadtxt("Sym10.txt", delimiter = ',').reshape([10, 10])
	
	def Calculate(self, C, derivatives):
		if 0 in derivatives:
			self.Value = - np.sum( C[0] * ( self.A @ C[0] ) )
		if 1 in derivatives:
			self.Gradient = [ - 2 * self.A @ C[0] ]

	def Hessian(self, V):
		return [ - 2 * self.A @ V[0] ]

class TestPrincipal(ut.TestCase):
	def __init__(self, *args):
		super().__init__(*args)
		self.Obj = Obj()
		self.Manifold = mv.Flag(np.eye(10)[:, :5]) # Initial guess
		self.Manifold.setBlockParameters([5])
		self.Tolerance = (1.e-5, 1.e-5, 1.e-5)
		self.Solution = np.linalg.eigh(self.Obj.A)[1][:, 5:]

	def testNewtonCG(self):
		M = mv.Iterate(self.Obj, [self.Manifold])
		tr = mv.TrustRegion()
		cg = mv.ConjugateGradient(M, 0, 1, (1e-4, 1e-4), M.getDimension(), 0)
		converged = mv.Newton(
				M, tr, cg, self.Tolerance, 8, 0
		)
		assert converged
		assert np.allclose(M.Ms[0].P @ M.Ms[0].P.T, self.Solution @ self.Solution.T, atol = 1e-5)

	def testNewtonMR(self):
		M = mv.Iterate(self.Obj, [self.Manifold])
		tr = mv.TrustRegion()
		mr = mv.MinRes(M, 0, 1, (1e-4, 1e-4), M.getDimension(), 0)
		converged = mv.Newton(
				M, tr, mr, self.Tolerance, 10, 0
		)
		assert converged
		assert np.allclose(M.Ms[0].P @ M.Ms[0].P.T, self.Solution @ self.Solution.T, atol = 1e-5)

	def testLBFGS(self):
		M = mv.Iterate(self.Obj, [self.Manifold])
		converged = mv.LBFGS(
				M, self.Tolerance,
				10, 43, 0.1, 0.75, 5, 0
		)
		assert converged
		assert np.allclose(M.Ms[0].P @ M.Ms[0].P.T, self.Solution @ self.Solution.T, atol = 1e-5)

	def testLanczos(self):
		M = mv.Iterate(self.Obj, [self.Manifold])
		M.setPoint([self.Solution], 1)
		M.Func.Calculate(M.getPoint(), [0, 1, 2])
		M.setGradient()
		Evals, Evecs = mv.Lanczos(M, M.getDimension(), 0, 0, 0)
		for i in range(len(Evecs)):
			residual = np.linalg.norm( M.ConstraintProjectedHessian(Evecs[i]) - Evals[i] * Evecs[i] )
			assert residual < 1e-5

if __name__ == "__main__":
	TestPrincipal().testNewtonCG()
	TestPrincipal().testNewtonMR()
	TestPrincipal().testLBFGS()
	TestPrincipal().testLanczos()

File: test/Principal.cpp
#include <Eigen/Dense>
#include <vector>
#include <iostream>
#include <Maniverse/Manifold/Flag.h>
#include <Maniverse/LinearSolver/ConjugateGradient.h>
#include <Maniverse/Optimizer/TruncatedNewton.h>
#include <Maniverse/Optimizer/LBFGS.h>
#include <Maniverse/Diagonalizer/Lanczos.h>

// Principal component analysis
// Finding the space spanned by the highest 5 eigenvectors
// Maximize L(C) = Tr[ C.t A C ]
// A \in Sym(10)
// C \in Flag(5; 10) = Gr(5; 10)

namespace mv = Maniverse;

class ObjPrincipal: public mv::Objective{ public:
	Eigen::MatrixXd A = Eigen::MatrixXd::Zero(10, 10);

	ObjPrincipal(){
		const double data[] = {
			#include "Sym10.txt"
		};
		std::memcpy(A.data(), &data, 10 * 10 * 8);
	};

	void Calculate(std::vector<Eigen::MatrixXd> C, std::vector<int> derivatives) override{
		if ( std::count(derivatives.begin(), derivatives.end(), 0) ){
			Value = - C[0].cwiseProduct( A * C[0] ).sum();
		}
		if ( std::count(derivatives.begin(), derivatives.end(), 1) ){
			Gradient = { - 2 * A * C[0] };
		}
	};

	std::vector<Eigen::MatrixXd> Hessian(std::vector<Eigen::MatrixXd> V) const override{
		return std::vector<Eigen::MatrixXd>{ - 2 * A * V[0] };
	};
};

#define __Check_Result__\
	std::cout << typeid(*this).name() << " " << __func__ << " ";\
	if ( converged ){\
		if ( ( M.Ms[0]->P * M.Ms[0]->P.transpose() - Solution * Solution.transpose() ).cwiseAbs().maxCoeff() < 1e-5 ){\
			std::cout << "\033[32mSuccess!\033[0m" << std::endl;\
		}else std::cout << "\033[31mFailed: Incorrect solution!\033[0m" << std::endl;\
	}else std::cout << "\033[31mFailed: Not converged!\033[0m" << std::endl;

#define __Check_Stability__\
	std::cout << typeid(*this).name() << " " << __func__ << " ";\
	for ( int i = 0; i < (int)Evecs.size(); i++ ){\
		const double residual = ( M.ConstraintProjectedHessian(Evecs[i]) - Evals[i] * Evecs[i] ).norm();\
		if ( residual > 1e-5 ) goto IncorrectCurvature;\
	}\
	std::cout << "\033[32mSuccess!\033[0m" << std::endl; return;\
	IncorrectCurvature: std::cout << "\033[31mFailed: Eigenvalue equation is violated!\033[0m" << std::endl;

class TestPrincipal{ public:
	ObjPrincipal Obj = ObjPrincipal();
	mv::Flag Manifold = mv::Flag(Eigen::MatrixXd::Identity(10, 5));
	std::tuple<double, double, double> Tolerance = {1.e-5, 1.e-5, 1.e-5};
	Eigen::MatrixXd Solution = Eigen::MatrixXd::Identity(10, 5);

	TestPrincipal(){
		Manifold.setBlockParameters({5});
		Eigen::SelfAdjointEigenSolver<Eigen::MatrixXd> es;
		es.compute(Obj.A);
		Solution = es.eigenvectors().rightCols(5);
	};

	void testTruncatedNewton(){
		mv::Iterate M(Obj, {Manifold.Share()});
		mv::TrustRegion tr;
		mv::ConjugateGradient cg(3, 1, {1e-4, 1e-4}, 1);
		const bool converged = mv::TruncatedNewton(
				M, tr, cg, Tolerance, 8, 1
		);
		__Check_Result__
	};

	void testLBFGS(){
		mv::Iterate M(Obj, {Manifold.Share()});
		const bool converged = mv::LBFGS(
				M, Tolerance,
				10, 43, 0.1, 0.75, 5, 1
		);
		__Check_Result__
	};

	void testLanczos(){
		mv::Iterate M(Obj, {Manifold.Share()});
		M.setPoint({Solution}, 1);
		M.Func->Calculate(M.getPoint(), {0, 1, 2});
		M.setGradient();
		const auto [Evals, Evecs] = mv::Lanczos(M, M.getDimension(), 1);
		__Check_Stability__
	};
};

int main(){
	TestPrincipal().testTruncatedNewton();
	TestPrincipal().testLBFGS();
	TestPrincipal().testLanczos();
}
